nemo/lightning/pytorch/callbacks/model_checkpoint.py

```diff
@@ -445,18 +471,70 @@ def _save_checkpoint(self, trainer: 'pytorch_lightning.Trainer', filepath: str)
                 self.deferred_ckpts_to_remove.append([])
             else:
                 storage_options = None
-            trainer.save_checkpoint(filepath, self.save_weights_only, storage_options=storage_options)
+            trainer.save_checkpoint(ckpt_filepath, self.save_weights_only, storage_options=storage_options)
 
-            ## NOTE: saving context happens synchronously always
-            from nemo.utils.get_rank import is_global_rank_zero
+            if self.always_save_artifacts and is_global_rank_zero():
+                TrainerContext.from_trainer(trainer).io_dump(ckpt_to_dir(filepath) / "artifacts")
 
-            if self.enable_nemo_ckpt_io and is_global_rank_zero():
-                TrainerContext.from_trainer(trainer).io_dump(ckpt_to_dir(filepath))
             if self.async_save:
                 logging.info(f'Scheduled async checkpoint save for {filepath}')
             else:
                 finalize_fn()
 
+    ## TODO: storage_options
+    def _drop_optimizer_states(self, trainer, filepath: Union[str, Path]) -> None:
+        # Get list of saved checkpoints
+        checkpoints = self._get_checkpoints_list(filepath)
+
+        checkpoint_index = len(checkpoints) - self.save_last_n_optim_states - 1
+        if len(checkpoints) > self.save_last_n_optim_states:
+            checkpoint_path = checkpoints[checkpoint_index]
+
+            logging.info(f"Loading '{checkpoint_path}' checkpoint to drop optimizer states...")
+            ## TODO: clean
+            checkpoint = trainer.strategy.load_checkpoint(Path(checkpoint_path) / ModelCheckpoint.MODEL_WEIGHTS_PATH)
+
+            # Remove the checkpoint version with optimizer states
+            self._remove_checkpoint(trainer, checkpoint_path)
+
+            checkpoint['optimizer'] = [None]
```

Comment:
this code can't live in ModelCheckpoint class because it's very specific to a MegatronStrategy.

I would suggest handling the "no optimizer" version in `strategy.save_checkpoint`

Reply:
I reverted this change. Support for `save_last_n_optim_states` will be added in a subsequent PR.